src/views/CreateShareStructure.vue

```diff
@@ -1,18 +1,346 @@
+import { ActionBindingIF, FormType, ShareClassIF } from '@/interfaces'
+
+// Components
+import { ListShareClass, ShareStructure } from '@/components/CreateShareStructure'
+
+@Component({
+  components: {
+    ListShareClass,
+    ShareStructure
+  }
+})
+export default class CreateShareStructure extends Vue {
+  // State
+  @State(state => state.stateModel.createShareStructureStep.shareClasses)
+  readonly shareClasses: ShareClassIF[]
+
+  // Actions
+  @Action setShareClasses!: ActionBindingIF
+  @Action setCreateShareStructureStepValidity!: ActionBindingIF
+
+  private newShareClass: ShareClassIF = {
+    id: null,
+    priority: null,
+    type: 'Class',
+    name: '',
+    hasMaximumShares: true,
+    maxNumberOfShares: null,
+    hasParValue: true,
+    parValue: null,
+    currency: 'CAD',
+    hasRightsOrRestrictions: false,
+    series: []
+  }
+
+  private newShareSeries: ShareClassIF = {
+    id: null,
+    priority: null,
+    type: 'Series',
+    name: '',
+    hasMaximumShares: true,
+    maxNumberOfShares: null,
+    hasParValue: true,
+    parValue: null,
+    currency: null,
+    hasRightsOrRestrictions: false
+  }
+
+  private sharesHelpSample: ShareClassIF[] = [{
+    'id': 1,
+    'priority': 0,
+    'type': 'Class',
+    'name': 'Sample Shares',
+    'hasMaximumShares': true,
+    'maxNumberOfShares': 1000,
+    'hasParValue': true,
+    'parValue': 1,
+    'currency': 'CAD',
+    'hasRightsOrRestrictions': true,
+    'series': []
+  }]
+
+  private showShareStructureForm: boolean = false
+  private addEditInProgress: boolean = false
+  private currentShareStructure: ShareClassIF | null = null
+
+  private activeIndex: number = -1
+  private parentIndex: number = -1
+  private nextId: number = -1
+  private helpToggle: boolean = false
+
+  // Lifecycle methods
+  mounted (): void {
+    this.setCreateShareStructureStepValidity(this.shareClasses.length > 0)
+  }
+
+  // Methods
+  private initNewShareClass (): void {
+    this.currentShareStructure = { ...this.newShareClass }
+    this.currentShareStructure.priority =
+    this.shareClasses.length === 0 ? 1 : this.shareClasses[this.shareClasses.length - 1].priority + 1
+    this.activeIndex = -1
+    this.parentIndex = -1
+    this.nextId = this.shareClasses.length === 0 ? 1 : (this.shareClasses.reduce(
+      (prev, current) => (prev.id > current.id) ? prev : current)).id + 1
+    this.addEditInProgress = true
+    this.showShareStructureForm = true
+  }
+
+  // Event Handlers
+  private initShareClassForEdit (index: number): void {
+    this.currentShareStructure = { ...this.shareClasses[index] }
+    this.activeIndex = index
+    this.parentIndex = -1
+    this.addEditInProgress = true
+    this.showShareStructureForm = true
+  }
+
+  private initNewShareSeries (shareClassIndex: number): void {
+    this.activeIndex = -1
+    this.parentIndex = shareClassIndex
+
+    let newList: ShareClassIF[] = [...this.shareClasses]
+    const parentShareClass = newList[shareClassIndex]
```

Comment:
It would be a good safety check to verify that shareClassIndex is valid before using it to deref an array.

Reply:
Ditto for index on line 180 above.

Reply:
Ditto for indices in all methods that use indices.

Reply:
I was looking at the flow of events and data and I dont think an invalid index will be passed back as the parent index / active index unless something goes wrong in reordering. In that case the fix would be on that side. I think we can leave it like this.